doc/design/blob-uuid-1.md

```diff
@@ -0,0 +1,45 @@
+A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<restofkey>`
+
+** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.
+
+**key generation function**: the current suggestion (from satra) for this function is UUID4.
+
+**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.
+
+## Upload (from a CLI)
+
+### Check process (for de-duplication):
+1. CLI sends expected digest with digest type (e.g., ETag)
+2. API checks for existence of ETag in identity table.
+3. if exists, returns key, if not returns an error.
+
+### Blob upload process:
+1. CLI initiates upload sending contentSize + ETag
```

Comment:
currently we do
```
            resp = self.post(
                "/uploads/initialize/",
                json={
                    "file_name": f"{dandiset_id}/{version_id}/{asset_path}",
                    "file_size": total_size,
                },
            )
```
so already conflating "pure blob" upload with an asset upload (has a file_name which could be different across blobs).  So I wonder if uploads should just be initiated from  `dandiset/.../versions/draft` end point?

Reply:
that is weird. i can see filename without all those other things being useful, since that can be put on the s3 object metadata for content-disposition, but what is the full file_name being used for in the upload process (dandiset/version seems completely unnecessary)

Reply:
`file_name` is actually unused by the API, it was required at some point but has been removed. Apparently it was never removed from the CLI, and having extra JSON fields doesn't cause any problems, the data is simply ignored. That line can be safely removed, it is meaningless.

Reply:
I think planning for different upload pipelines at this point is not productive. For now we have a well defined problem, uploading a file from a client machine using the CLI. If we want to support different ways to ingest files in the future, we can build a different upload flow at that time.

My understanding is that to get to release faster, we were going to just trust the CLI that content is valid for now. Handling metadata and validating during initialization would make things substantially more complicated. Is this a migration blocker, or can we leave it out for now @yarikoptic ?

Reply:
i think @yarikoptic - simply meant sending the info in the json header in that post (and not really full metadata validation). but i can also see sending contentsize/digest info in the call to finalize upload, since that's where validation would happen. 

and i do agree that making that json match with the subset of assetmetadata keys would probably be a good thing, but not essential at this point.

so the payload on `/uploads/validate` the upload would be something like this.

```json
{ 
  "identifier": uuid,
  "contentSize":  12345,
  "digest":  {"dandi:dandi-s3-etag": "aadd-123"}
}
```

this formulation still needs a change in the cli.

Reply:
I think `identifier` not be included, or at least be optional:
```
{ 
  "contentSize":  12345,
  "digest":  {"dandi:dandi-s3-etag": "aadd-123"}
}
```
When the CLI needs to upload file, the first step is to calculate some digest (in this case ETag) and ask the API using this endpoint if it already exists; there's no way for the CLI to know the UUID.
Since we already need to support that use case, I don't see what specifying `identifier` gains us.

Reply:
FWIW re `file_path` and content-disposition: I would stand that blobs have no filename (IMHO UUID is better than a wrong/misleading/disclosing one from the first ever uploaded copy of the blob):
- Pre-signed urls produced from `*/assets/{uuid}/download` endpoint can have appropriate for the asset content-disposition
- git-annex doesn't care
- "published" manifests would have proper paths and thus filenames for each of the blobs

So indeed, no need to provide file path or name anywhere here

Reply:
> Pre-signed urls produced from */assets/{uuid}/download endpoint can have appropriate for the asset content-disposition

has this been tested. let me add it as an issue.

Reply:
@yarikoptic I agree with returning the blob ID only from `/uploads/validation/`, but I don't see the need for an intermediate `upload-id`. To support deduplication, `/uploads/validate/` needs to accept a request body like:
```
{ 
  "contentSize":  12345,
  "digest":  {"dandi:dandi-s3-etag": "aadd-123"}
}
```
Since the ETag is already there (and is already being relied on to uniquely identify blobs), we can just look up the upload using that. I don't see what including an `upload-id` gains us.

Reply:
what if two people are uploading the same blob?

Reply:
would you randomly choose the first uuid?

Reply:
To have a collision, the second person would need to call `/uploads/initialize/` with the same ETag as the first person. This request should fail because there is already an upload.

Reply:
i thought we allowed for parallel uploads of the same content (since we are using UUID, and the ETag check does not happen on the upload table) - which if we check with ETag during upload stage while nice, would make the person A wait for person B to finish uploading. this would also ensure that the fastest uploaded blob gains precedence.

Reply:
> I don't see what including an upload-id gains us

- a clear identification of the upload in question, and avoidance of matching by metadata provided in `/upload/initialize/` and thus avoiding problems with duplicate uploads as @satra has mentioned - we can just allow them to happen and the first one to complete would be the one the `key` of which would be used for that blob
- anything about upload (cancel etc) should be done via its `upload_id` instead of matching uploads and possibly canceling some other parallel one (ATM we do not have `DELETE` on uploads but I guess we should get one eventually)
That is why below I would prefer [my](https://github.com/dandi/dandi-api/pull/150#discussion_r591868008) over [@dchiquito ](https://github.com/dandi/dandi-api/pull/150#discussion_r591902088) proposal below.

Reply:
OK, I think at this point we're really overloading `/upload/initialize/`. How about breaking it into two endpoints:
 * `GET /blobs/` checks for an existing asset with an arbitrary digest (but ETag for now):
 ```
 { 
  "contentSize":  12345,
  "digest":  {"dandi:dandi-s3-etag": "aadd-123"}
}
```
* `POST /uploads/validate/{upload_id}/` starts validation for an upload, identified by upload ID.